gcs.go

```diff
@@ -109,49 +123,98 @@ func (s GCStore) GetChunk(id ChunkID) (*Chunk, error) {
 
 // StoreChunk adds a new chunk to the store
 func (s GCStore) StoreChunk(chunk *Chunk) error {
+
 	ctx := context.TODO()
 	contentType := "application/zstd"
 	name := s.nameFromID(chunk.ID())
+
 	var (
 		b   []byte
 		err error
+		log = Log.WithFields(logrus.Fields{
+			"bucket": s.bucket,
+			"name":   name,
+		})
 	)
+
 	if s.opt.Uncompressed {
 		b, err = chunk.Uncompressed()
 	} else {
 		b, err = chunk.Compressed()
 	}
+
 	if err != nil {
+		log.WithError(err).Error("Cannot retrieve chunk data")
 		return errors.Wrap(err, s.String())
 	}
+
 	r := bytes.NewReader(b)
 	w := s.client.Object(name).NewWriter(ctx)
 	w.ContentType = contentType
 	_, err = io.Copy(w, r)
+
 	if err != nil {
+		log.WithError(err).Error("Error when copying data from local filesystem to object in GCS bucket")
 		return errors.Wrap(err, s.String())
 	}
 	err = w.Close()
 	if err != nil {
+		log.WithError(err).Error("Error when finalizing copying of data from local filesystem to object in GCS bucket")
 		return errors.Wrap(err, s.String())
 	}
+
+	log.Debug("Uploaded chunk to GCS bucket")
+
 	return nil
 }
 
 // HasChunk returns true if the chunk is in the store
 func (s GCStore) HasChunk(id ChunkID) (bool, error) {
+
 	ctx := context.TODO()
 	name := s.nameFromID(id)
+
+	var (
+		log = Log.WithFields(logrus.Fields{
+			"bucket": s.bucket,
+			"name":   name,
+		})
+	)
+
 	_, err := s.client.Object(name).Attrs(ctx)
-	return err == nil, nil
+
+	if err == storage.ErrObjectNotExist {
+		log.WithField("exists", false).Debug("Chunk does not exist in GCS bucket")
+		return false, nil
+	} else if err != nil {
+		log.WithError(err).Error("Unable to query attributes for object in GCS bucket")
+		return false, err
+	} else {
+		log.WithField("exists", true).Debug("Chunk exists in GCS bucket")
+		return true, nil
+	}
 }
 
 // RemoveChunk deletes a chunk, typically an invalid one, from the filesystem.
 // Used when verifying and repairing caches.
 func (s GCStore) RemoveChunk(id ChunkID) error {
 	ctx := context.TODO()
 	name := s.nameFromID(id)
-	return s.client.Object(name).Delete(ctx)
+
+	var (
+		log = Log.WithFields(logrus.Fields{
+			"bucket": s.bucket,
+			"name":   name,
+		})
+	)
+
+	err := s.client.Object(name).Delete(ctx)
```

Comment:
Could cause a confusing error since it'll print both, the error message and the success message. Perhaps something more explicit would work.

```go
if err := s.client.Object(name).Delete(ctx); err != nil {
  log.WithError(err).Error("Unable to delete object in GCS bucket")
  return err
}
log.Debug("Removed chunk from GCS bucket")
return nil
```

Reply:
Good catch, thanks. Flow changed to something close to your suggestion.